src/adr/0019-distributed-blob-garbage-collector.md

```diff
@@ -0,0 +1,432 @@
+# 19. Distributed blob garbage collector
+
+Date: 2020-02-18
+
+## Status
+
+Proposed
+
+## Context
+
+The body, headers, attachments of the mails are stored as blob on a blob store.
+In order to save space on the those stores, those blobs are de-duplicated using an hash of their content.
+This makes the safe deletion of one of those blob a non trivial problem as we can't delete one blob without ensuring
+that all references to it are themselves deleted.
+
+## Decision
+
+To address this issue, we propose to implement a distributed blob garbage collector build upon the previously developed
+Distributed Task Manager.
+This garbage collector will keep track of the references pointing toward a blob in a `References` table.
+It will also keep track of the deletion requests for a blob in a `Deletions` table.
+When the algorithm runs it will fetch in the `Deletions` table the blobs considered to be effectively deleted,
+and will check in the `References` table if there is still some references to them. If there is no more reference to a blob, 
+it will be effectively deleted from the blob store.
+
+To avoid concurrency issue, where we could garbage collect a blob at the same time a new reference to it appears.
+A `reference generation` notion will be added. At a given interval a new `reference generation` will be emitted,
+since then all new blobs will point to this new generation. And their de-duplicating id which before where constructed
+using only the hash of their content will now include this `reference generation` too.
+
+So a `garbage collection iteration` will run only on the `reference generation` `n-1` to avoid concurrency issue.
+
+
+## Alternatives
+
+Not de-duplicating the blobs content, this simple approach was not kept because in some cases it involves storing the same blob a lot of times, which is really slow and costly.
```

Comment:
In my opinion, we should have both. With the new BlobStore interface, the de-duplication is no longer tied to a backend so we should be able to use backends with and without de-duplication.

Reply:
What would be the rational not to use de-duplication ?

Reply:
* More performance overall
* Data security
* Single source of truth (de-dup means a database + a blob store)
* Complexity of operation + cpu/io cost vs more blobstore storage cost

And I'm sure the list goes on